platform-sdk/docs/proposals/DP-0x/0X - Tss block signing - Cryptography APIs.md

```diff
@@ -0,0 +1,661 @@
+3. **swirlds-cryptography-tss**: Gradle module for the TSS Library. This library-only client is the consensus node so that it will remain close to it in the `hedera-services` repository under the `platform-sdk` folder.
+4. **swirlds-cryptography-signatures**: Gradle module for the Bilinear Pairings Signature Library.
+5. **swirlds-cryptography-pairings-api**: Gradle module for the Bilinear Pairings API. Minimizes the impact of adding or removing implementations.
+6. **swirlds-cryptography-alt128**: Gradle module that will implement the Bilinear Pairings API using alt-128 elliptic curve. That curve has been chosen following EVM's support. Support for that curve will be provided by arkworks rust library. The module will include Java and Rust codes that will be compiled for all possible system architectures and distributed in a jar under a predefined structure.
+
+### Libraries Specifications
+
+#### Swirlds Crypto TSS Library
+##### Overview
+This library implements the Groth21 TSS-specific primitives.
+
+##### Constraints
+Once built, this library will depend on artifacts constructed in other repositories.
+
+##### Public API
+
+##### Example
+###### 1. Bootstrapping
+```java
+   TssService service = new TssService(signatureScheme);
+   PairingPrivateKey persistentParticipantKey = Requirements.loadECPrivateKey();
+   List<PairingPublicKey> persistentPublicKeys = Requirements.loadECPublicKeys();
+   TssParticipantDirectory directory = 
+        service.participantDirectoryBuilder()
+           .self(/**/ 0, persistentParticipantKey)
+           .withClaim(0, 5, persistentPublicKeys.get(0))
+           .withClaim(1, 2, persistentPublicKeys.get(1))
+           .withClaim(2, 1, persistentPublicKeys.get(2))
+           .withClaim(3, 1, persistentPublicKeys.get(3))
+          .build();
+   
+   //One can then query the directory
+   int n = directory.getTotalNumberOfShares();
+   List<TssShareId> privateShares = directory.getOwnedShares();
+   List<TssShareId> shareIds = directory.getShareIds();
+```
+###### 1. Create TssMessage
+```java
+
+   TssSecretShare s = directory.createRandomSecretShare();
+   TssMessage m = service.createTssMessage(directory, s, threshold);
+```
+
+###### 2. Validation of TssMessage
+```java
+   List<TssMessage> messages = Requirements.receiveTssMessages();
+   for(TssMessage m : messages){
+     m.validate(directory);
+   }
+```
+###### 3. Processing of TssMessage
+```java
+   List<TssMessage> messages = Requirements.receiveTssMessages();
+   TssService cachedService = service.withCache(size);
+   
+   TssProcessingResult result = cachedService.processMessages(directory, messages);
+   result.isValid();
+   TssParticipantDirectory updatedDirectory = result.getDirectory();
+```
+######  Sign
+```java
+   byte[] message = Requirements.messageToSign();
+
+   List<TssSignature> signatures = updatedDirectory.produceSignatures(updatedDirectory, message);
```

Comment:
TssDirectory contains data and access to it.
TssServices contains operations. Uses a directory.
At least in theory.

Reply:
I'm not convinced that it is a good idea to bundle all of the data up into a single object.

The platform HATES shared data structures. If we bundle everything up, that means the platform will need to coordinate access across many components.

The platform operation proposal that I had drafted utilizes maps and lists for this reason.

Reply:
I understand.
Let's say we separate them again into multiple data structures.
Where would all those references live?

Reply:
Do you mean from the context of the platform?

Reply:
Exactly. 
Where would we keep track of all those lists and map references, such as privateSharesList, publicSharesMap, and shareClaims?
What structure would hold the references to be accessible to the platform code?

My goal is to reduce the number of objects the user needs to know how to create to operate with the library. 
In the draft proposal, users need to: 
 * create ShareClaims to create TssMessages.
 * use claims and EcKeys to validate TssMessages.
 * use a list of TssMessages and a private key and claims to create private shares.
 * use a list of TssMessages, public keys, and claims to create public shares.
 * aggregate public shares to produce the ledger.
 * use public shares to validate received signatures
 * use private shares to create partial signatures.
 * use partial signatures to create aggregate signatures.

Sometimes, they need to interact with a manager object to invoke the desired operation; sometimes, the operation resides in the object itself.

I like that because it is 100% object-oriented, but all those interactions are complex to specify, document, and use.

Having a centralized piece where one can check all operations exposed by the library would make things easier for the user (IMO).
If we move to that approach, returned objects become sillier. They are returned to the user, who then returns them to the library. 
Having fewer objects as parameters would help with the interaction between the user and the library.

I am trying to think how can we achieve the two things:
  Simplified interactions.
  Structure independence to avoid the need for synchronization.

Reply:
The way that the platform would use the various elements is captured in the wiring diagram in this doc https://www.notion.so/swirldslabs/TSS-Platform-Architecture-04b15df371ba4b1d848360542e05a030

The various platform components are responsible for maintaining an internal state, based on the tasks they each perform.